domain.c

```diff
@@ -874,78 +985,79 @@ int domain_nonrecursively_combine_topTree(struct local_topnode_data * topNodes)
  *  in pieces of eight segments until each segment holds at most a certain
  *  number of particles.
  */
-int domain_determineTopTree(struct local_topnode_data * topNodes)
+int domain_determineTopTree(struct local_topnode_data * topTree)
 {
     int i, j, sub;
     int errflag, errsum;
 
-    int64_t totgravcost, gravcost = 0;
+    struct local_particle_data * LP = (struct local_particle_data*) mymalloc("LocalParticleData", NumPart * sizeof(LP[0]));
 
-#pragma omp parallel for reduction(+: gravcost)
+#pragma omp parallel for
     for(i = 0; i < NumPart; i++)
     {
-        int64_t costfac = domain_particle_costfactor(i);
-        gravcost += costfac;
+        LP[i].Key = P[i].Key;
+        LP[i].Cost = domain_particle_costfactor(i);
     }
 
-    /*We need TotNumPart to be up to date*/
-    sumup_large_ints(1, &NumPart, &TotNumPart);
-    MPI_Allreduce(&gravcost, &totgravcost, 1, MPI_INT64, MPI_SUM, MPI_COMM_WORLD);
+    walltime_measure("/Domain/DetermineTopTree/Misc");
 
-    int64_t costlimit, countlimit;
+    /* Watchout : must disgard proximity of particle type; this ordering is only required by LocalRefine */
 
-    costlimit = totgravcost / (All.TopNodeIncreaseFactor * All.DomainOverDecompositionFactor * NTask);
-    countlimit = TotNumPart / (All.TopNodeIncreaseFactor * All.DomainOverDecompositionFactor * NTask);
+    mpsort_mpi(LP, NumPart, sizeof(struct local_particle_data), mp_order_by_key, 8, NULL, MPI_COMM_WORLD);
```

Comment:
The sorting has nothing to do with subsampling. It is required by the faster tree build algorithm. Global sorting enhances the accuracy of the final global topTree. Local sorting also produces correct results (but less accurate). I added comments to clarify this.